Contact Form Gutenblock

georgestephanis opened the pull request:
GM2018 Work In Progress.

Do not merge!

This is built using the https://github.com/Automattic/wp-calypso/ build tools -- something like this:

```sh
#!/bin/bash

CALYPSO_DIR="/Users/georgestephanis/code/wp-calypso/"
JETPACK_DIR="/Users/georgestephanis/vagrant-local/www/wordpress-develop/public_html/build/wp-content/plugins/jetpack/"

cd $CALYPSO_DIR

cp -rf ${JETPACK_DIR}modules/contact-form/block/* ${CALYPSO_DIR}client/gutenberg/extensions/contact-form/
npm run sdk -- gutenberg ${CALYPSO_DIR}client/gutenberg/extensions/contact-form/
cp -rf ${CALYPSO_DIR}client/gutenberg/extensions/contact-form/build/* ${JETPACK_DIR}modules/contact-form/block/build/
```

but for testing purposes, the compiled js file is bundled in the pr.

## TBD ##

- [x] Make the save button text editable.
- [ ] Rendering on the front-end (waiting on externals -- @dmsnell )
- [x] Deletion of multi fields
- [x] More transforms
- [ ] ???

## Testing instructions ##

Clone the branch, and add a `contact form` block in Gutenberg.

This is also mirrored to Calypso in https://github.com/Automattic/wp-calypso/pull/27688

## Proposed changelog entry ##

Add a Gutenberg Block for the Jetpack Contact Form

georgestephanis committed:
Ignore node_modules regardless of depth

georgestephanis committed:
Add build tools for contact form

Easier than messing around with outdated gulp config in master

georgestephanis committed:
First pass.

georgestephanis committed:
Redo fields.

georgestephanis committed:
Update some icons, renders, etc.

georgestephanis committed:
Correct how I'm grabbing the label.

georgestephanis committed:
Assorted changes.

georgestephanis committed:
d'oh copypasta fail

georgestephanis committed:
Do it the easy way, not the tidy way.

It works.  Don't knock it.

georgestephanis committed:
Begin saving `to` and `subject` fields, add `required` toggle.

georgestephanis committed:
whitespace tidying / reorg

georgestephanis committed:
Abstract out the label stuff for simpler components and reusability.

georgestephanis committed:
Add the compiled version to make for easier testing.

georgestephanis committed:
Start experimenting with options parameters and rendering.

georgestephanis committed:
Have each option save its value.

georgestephanis committed:
compile!

georgestephanis committed:
tabs

georgestephanis committed:
whitespace tidying

georgestephanis committed:
update build file

georgestephanis committed:
Stub out render functions

georgestephanis committed:
Get radio control working

Saves correctly.

georgestephanis committed:
Get Select fields running, and make "Add New" translateable

georgestephanis committed:
Compile!

georgestephanis committed:
Add more php parsing for varied block types.

Maps to existing shortcode parsers.

georgestephanis committed:
Handle arrays for atts as gutenberg uses json

georgestephanis committed:
whitespace tidying

georgestephanis committed:
We don't need this for numeric attrs, as gutenberg won't make it numeric

georgestephanis committed:
add a bool handler for required

georgestephanis committed:
Consolidate the multi select options

georgestephanis committed:
Add some rudimentary styles to the back-end form editor.

georgestephanis committed:
build

georgestephanis committed:
Tidy up legacy supportHTML option to object and change widgets to common

mapk commented:
I have some questions around a11y and usability while editing fields within the form's block. @MichaelArestad 

Currently, @georgestephanis has the fields show a border on hover and then highlight on focus. This is obvious in a great way.

![edit-borders](https://user-images.githubusercontent.com/617986/46547838-1ba67e80-c882-11e8-98d9-a1b5fd403cc6.gif)

Can we get away with lessoning the amount of UI? Maybe something like this which only shows the focus highlight?

![edit-a11y-focus](https://user-images.githubusercontent.com/617986/46547813-029dcd80-c882-11e8-8a10-4eef5d22ae34.gif)

Here's a concept of no borders at all... maybe this isn't not enough?

![edit-no-borders](https://user-images.githubusercontent.com/617986/46547848-23662300-c882-11e8-927e-bb408a72a60a.gif)

mapk committed:
Minor styling changes for labels and spacing.

mapk committed:
Merge branch 'gm18/grunionblock' of github.com:Automattic/jetpack into gm18/grunionblock

georgestephanis committed:
Purge some of the local build tools

As I prepare to use Calypso to build the jsx instead.

georgestephanis committed:
Start experimenting with Block Transforms.

georgestephanis committed:
Merge branch 'gm18/grunionblock' of https://github.com/Automattic/jetpack into gm18/grunionblock

georgestephanis committed:
Change methodology to use Calypso build tool SDK?

Kinda similar to https://github.com/Automattic/wp-calypso/pull/27386

georgestephanis committed:
Build via calypso builder instead.

Commands used:

```bash
cp ~/vagrant-local/www/wordpress-develop/public_html/build/wp-content/plugins/jetpack/modules/contact-form/js/gutenblock.jsx /Users/georgestephanis/code/wp-calypso/client/gutenberg/extensions/contact-form/editor.js
cp ~/vagrant-local/www/wordpress-develop/public_html/build/wp-content/plugins/jetpack/modules/contact-form/css/gutenblock.css /Users/georgestephanis/code/wp-calypso/client/gutenberg/extensions/contact-form/editor.scss
npm run sdk -- gutenberg ~/code/wp-calypso/client/gutenberg/extensions/contact-form/
cp ~/code/wp-calypso/client/gutenberg/extensions/contact-form/build/editor.js ~/vagrant-local/www/wordpress-develop/public_html/build/wp-content/plugins/jetpack/modules/contact-form/js/gutenblock.js
```

georgestephanis commented:
Building is now done via a checkout of https://github.com/Automattic/wp-calypso/ instead of locally in this repo -- props @gwwar for running me through it last week.

```sh
cp ~/vagrant-local/www/wordpress-develop/public_html/build/wp-content/plugins/jetpack/modules/contact-form/js/gutenblock.jsx /Users/georgestephanis/code/wp-calypso/client/gutenberg/extensions/contact-form/editor.js
cp ~/vagrant-local/www/wordpress-develop/public_html/build/wp-content/plugins/jetpack/modules/contact-form/css/gutenblock.css /Users/georgestephanis/code/wp-calypso/client/gutenberg/extensions/contact-form/editor.scss
npm run sdk -- gutenberg ~/code/wp-calypso/client/gutenberg/extensions/contact-form/
cp ~/code/wp-calypso/client/gutenberg/extensions/contact-form/build/editor.js ~/vagrant-local/www/wordpress-develop/public_html/build/wp-content/plugins/jetpack/modules/contact-form/js/gutenblock.js
```

I'd love to see an easier workflow to do it like that, but hey -- it works.

gwwar commented:
If folks have a Calypso branch, I believe jetpack block artifacts are built in a new circle ci workflow. https://circleci.com/gh/Automattic/wp-calypso/110207#artifacts/containers/0 So you don't need to run the commands yourself, though we do need to use the circle apis to download them. I think workflows can definitely be tweaked as we run into pain points.

@sirreal or other folks on Luna may have more info on this.

![screen shot 2018-10-08 at 10 38 24 am](https://user-images.githubusercontent.com/1270189/46624667-6294ae00-cae6-11e8-9d06-4a1e82465e82.png)

![screen shot 2018-10-08 at 10 38 44 am](https://user-images.githubusercontent.com/1270189/46624663-5e689080-cae6-11e8-881d-fbbe7f2e8ca5.png)

georgestephanis commented:
@gwwar Yeah, I'm trying to save the editor.js and editor.scss to a new folder and make a pr for calypso from them, but there's some oopsies happening so I'm just using it as a passthrough at the moment.

What I'd like to be able to do is something like -- `npm run sdk -- gutenberg ~/path/to/whatever.jsx -o ~/path/to/whatever.js` and ideally just have it transpile that file directly from the `.jsx` to the `.js` file in the same directory, outside of Calypso's repository entirely.

georgestephanis committed:
spaces 👉🏼  tabs 🆒

gwwar commented:
The sdk does has output directory options and a watch command. 

@dmsnell should update the [readme](https://github.com/Automattic/wp-calypso/blob/master/docs/sdk.md).

so `-o or --output-dir` and `--watch`

https://github.com/Automattic/wp-calypso/blob/d911b7c4f460bbd5e742060ec6ad411380af0cd5/bin/sdk-cli.js#L87

georgestephanis commented:
Yeah, but it requires the source file to be `editor.js` -- Ideally I'd like to be able to specify the source file manually, even if it's named `somethingelse.jsx` -- for visual designation of source / transpiled files I've preferred to use `.jsx`/`.js` kinda akin to `.scss`/`.css`

gwwar commented:
I'm not super familiar with the workflow Jetpack is using at the moment, maybe @Automattic/calypso-sdk can chime in, but would we not rebuild the full Jetpack preset?

I think folks picked convention over configuration in https://github.com/Automattic/wp-calypso/pull/27257, which seems fine to me, especially since Mobile needs this for core blocks so they can override pieces for use in React Native.

georgestephanis committed:
Remove map file no longer generated by calypso

georgestephanis committed:
Make the 'submit' button editable.

MichaelArestad commented:
> Can we get away with lessoning the amount of UI? Maybe something like this which only shows the focus highlight?

@mapk absolutely. I will audit and make recommendations to the styles this week.

georgestephanis committed:
Tidy up some code standards stuff

jeherve commented:
Closing in favor of https://github.com/Automattic/wp-calypso/pull/27688

georgestephanis commented:
Reopening this after chatting briefly with @sirreal in Slack.

This PR has a number of stuff that the Calypso PR can't -- namely all the PHP changes, as well as is easier for me to do the primary development in and has the full history.  The Calypso branch was just a place to test and share reimplementing it in Calypso and to get commit checks running against it.

georgestephanis committed:
Make it possible to swap the button text, and some whitespace tidying.

georgestephanis committed:
Tidy up how we're passing around functions to set attributes.

georgestephanis committed:
Whitespace

georgestephanis committed:
Ignore the cform gutenblock while I'm working on it.

georgestephanis committed:
Whitespace tidying and change option add button.

georgestephanis committed:
Update the styling of some elements.

georgestephanis committed:
Enhancement: add ability to delete options from multi select field types

georgestephanis committed:
Add placeholder to multi fields

georgestephanis committed:
Add alternate transforms.

georgestephanis committed:
whitespace formatting / tidying

georgestephanis committed:
Switch  to

georgestephanis committed:
Build.

georgestephanis committed:
Add matches and transforms for single/multi field types

georgestephanis commented on modules/contact-form/js/gutenblock.jsx:
```diff
@@ -0,0 +1,568 @@
+					<TextControl
+						className="button button-primary button-default grunion-submit-button"
+						value={ this.props.submit_button_text ? this.props.submit_button_text : __( 'Submit »' ) }
+						onChange={ ( x )=>this.props.setAttributes( { submit_button_text: x } ) }
```
I've been getting some precommit warnings about using this method for setting the onchange -- do we have any recommendations for a better way to do this?

coderkevin replied:
I'm guessing it's because of the anonymous function, right? It's mostly for performance reasons due to unnecessary re-rendering. Better to move it to a function defined in the class. Here's a decent explanation: http://johnnyji.me/react/2016/06/27/why-arrow-functions-murder-react-performance.html

georgestephanis commented on modules/contact-form/js/gutenblock.jsx:
```diff
@@ -0,0 +1,568 @@
+										this.props.setAttributes( {
+											options: _.pluck( $options.toArray(), 'value' )
+										} );
+									}.bind( this ) }
```
I've also gotten some warnings about the `.bind()` here -- any recommendations on how to rewrite it instead?

blowery replied:
the bind here is fine. The warnings are related to possible perf problems with passing a new function instance to the component on each render, which can cause the component to re-render unnecessarily.

If you really want to rewrite it, I'd do two things: Make up a new GrunionOption component that takes the option and an onChange callback as props. Then in the GrunionOption component, it would register the onChange handler against a locally bound function and call this.props.onChange, passing out the option that was passed in.  The option component would look something like:
```
class GrunionOption extends React.Component {
  handleOnChange = event => {
    this.props.onChange( this.props.option )
  }

  render() {
    ...
    return <input onChange={ this.handleChange } ... />
  }
}

coderkevin replied:
Same reasoning as above. The creation of a new function each time this renders causes the subcomponent to no longer match its props and forces it to re-render to. The bind does the same. I'd make an `onChange` function outside of render and pre-bind it in the constructor (or use arrow function and it binds automatically.)

georgestephanis commented on modules/contact-form/js/gutenblock.jsx:
```diff
@@ -0,0 +1,568 @@
+/*global jQuery _ wp*/
```
I seem to recall someone remarking that underscore isn't available in Calypso -- I'm currently using `_.map()` `_.pluck()` `_.clone()` and `_.defaults()` -- are there recommended ways to abstract these out of either underscore or lodash or to make it work equally well with either?

Is `jQuery` available in Calypso?  I'm currently using `jQuery.closest()` and `jQuery.find()` to do some traversal, but can probably rewrite it in vanilla javascript if necessary.

coderkevin replied:
Lodash is used in Calypso, but typically with import syntax. (e.g. `import { pluck } from 'lodash' )
We've really tried hard not to use jQuery in Calypso at all, so I'd recommend against it if you can. If you're trying to inspect or manipulate the DOM using jQuery, you probably need to be careful doing this as it can really interfere with the way React works.

blowery replied:
Yeah, no underscore in Calypso, only lodash. They're sometimes compatible, but I wouldn't rely on it.

jQuery isn't available globally in Calypso. Gutenberg pulls in v3 for some things now, but they're working on removing it.

georgestephanis reviewed (commented):
The bits I'm most uncertain of are highlighted here in this review -- happy to make further changes for Calypso compatability, just let me know what is and isn't problematic.

Also a bit of help on the onchange issues I highlighted would be appreciated -- I believe the issue with the current method is that it winds up redeclaring the functions and can result in a memory issue eventually?

coderkevin commented on modules/contact-form/js/gutenblock.jsx:
```diff
@@ -0,0 +1,568 @@
+}
+
+class GrunionFieldMultiple extends Component {
+	render() {
```
This render function is really long and confusing, consider breaking it up?

coderkevin commented on modules/contact-form/js/gutenblock.jsx:
```diff
@@ -0,0 +1,568 @@
+									value={ option }
+									placeholder={ __( 'Enter your option value here…' ) }
+									onChange={ function( x ) {
+										const $options = jQuery( x.target ).closest( 'ol' ).find( 'input.option' );
```
jQuery and React aren't terribly compatible. I'd try to do this without jQuery. Let me know if you'd like to discuss for more options so I can better understand what you're trying to accomplish here.

coderkevin commented on modules/contact-form/js/gutenblock.jsx:
```diff
@@ -0,0 +1,568 @@
+/*global jQuery _ wp*/
```
This file is really big and is doing a whole lot of stuff. I'd recommend breaking it into several files to make it easier to traverse, manage, and test.

georgestephanis replied:
Probably agreed.  For the moment it's just grown as I was writing it and I'd love to break it up in a bit -- may swipe you for some help with that.

coderkevin commented on modules/contact-form/js/gutenblock.jsx:
```diff
@@ -0,0 +1,568 @@
+	}
+}
+
+class GrunionFieldLabel extends Component {
```
Many of these smaller components don't need to be a class. They can just be a function. Components really only need to be a class if they need lifecycle functions (e.g. `componentDidMount`), or if they need to define bound event handler functions (and sometimes not even then)

georgestephanis replied:
I've changed them over to functions where possible.

georgestephanis committed:
Abstract out the processing logic.

This is a perf win and clears some static code analysis issues.

georgestephanis committed:
build

georgestephanis committed:
Punt jQuery.

georgestephanis committed:
Stop using underscore for `_.map()`

georgestephanis committed:
Vanilla js way of cloning an object in js.

Props @darkwing for the idea!

https://davidwalsh.name/javascript-clone-array

georgestephanis committed:
Remove last vestiges of underscore.

georgestephanis committed:
build

georgestephanis committed:
Simplify some classes down to functions.

georgestephanis committed:
Simplify out how we're adding new items

georgestephanis committed:
Add wp global back in -- oops

georgestephanis committed:
Pull out setAttributes fn call.

georgestephanis committed:
Abstract out onChanges

georgestephanis committed:
Switch to placeholders

georgestephanis committed:
tabs > spaces

georgestephanis committed:
Build

simison commented on modules/contact-form/css/gutenblock.css:
```diff
@@ -0,0 +1,84 @@
+}
+.editor-block-list__block[data-type="grunion/field-radio"] ol li:before {
+	content: '🔘 ';
+}
```
This seems like something we shouldn't need to do — what was the reasoning of using CSS `content` instead of having them inline at HTML?

(I didn't yet try running the block so I might be missing something)

georgestephanis replied:
Using css means that we can have the same Component rendering the markup for Multiple Checkboxes, Radio Buttons, and Select Boxes -- just using styling to make them visually representative of their outputs.

simison commented on modules/contact-form/css/gutenblock.css:
```diff
@@ -0,0 +1,84 @@
+	margin: 0;
+}
+
+span.required {
```
This one looks like it needs `grunion-` prefix as it would end up being global and affect things outside the block.

georgestephanis replied:
That's fair.

simison commented on modules/contact-form/css/gutenblock.css:
```diff
@@ -0,0 +1,84 @@
+
```
Suppose this file is CSS instead of SCSS just to make building easier for now?

Let's switch it to SCSS as soon as we can so that we can rely on nesting, colour variables and other things like that so that as the CSS grows, we don't need to do too much refactoring.

georgestephanis replied:
Yarp, that's the plan.

simison commented on modules/contact-form/js/gutenblock.jsx:
```diff
@@ -0,0 +1,630 @@
+ * Block Registrations:
+ */
+
+registerBlockType( 'grunion/form', {
```
Let's stick to the same namespace as the rest of the Jetpack blocks use? `a8c/form` 

Another discussion if it should be `jetpack/form` instead but whatever the name, consistency is the way to go. 👌

georgestephanis replied:
Yarp, I need to do a pass renamespacing it soon.  Grunion was just the legacy handle of the module.

simison commented on modules/contact-form/js/gutenblock.jsx:
```diff
@@ -0,0 +1,630 @@
+						/>
+					</PanelBody>
+				</InspectorControls>
+				<div className="grunion-form">
```
Should we rely on `props.className` generated by Gutenberg here instead?

georgestephanis replied:
May as well append it in as well, no need to have it be an either/or.

simison commented on modules/contact-form/js/gutenblock.jsx:
```diff
@@ -0,0 +1,630 @@
+				</InspectorControls>
+				<div className="grunion-form">
+					{this.props.children}
+					<TextControl
```
Should we use Gutenberg's [Button](https://github.com/WordPress/gutenberg/tree/a9a27bb25f54b0d61b797006c21bc20b4685dbe1/packages/components/src/button) component here instead?

georgestephanis replied:
Can Button be used to allow the text of it to be editable as TextControl can?

simison replied:
Sounds like there's a good reason to keep it as `TextControl`! Unless HTML attribute `contenteditable` somehow works but seems hacky.

simison commented on modules/contact-form/js/gutenblock.jsx:
```diff
@@ -0,0 +1,630 @@
+					<TextControl
+						className="button button-primary button-default grunion-submit-button"
+						value={ this.props.submit_button_text }
+						placeholder={ __( 'Submit »' ) }
```
`»` feels extraneous, could we leave it out?

georgestephanis replied:
That's the legacy string that it already uses.  It didn't feel worth changing it and affecting translations.

simison replied:
Looks like it's already included in translations so we'd be good to do the change:

https://translate.wordpress.org/projects/wp-plugins/jetpack/dev/en-gb/default?filters%5Bstatus%5D=either&filters%5Boriginal_id%5D=600163&filters%5Btranslation_id%5D=29847455

simison commented on modules/contact-form/js/gutenblock.jsx:
```diff
@@ -0,0 +1,630 @@
+					className='grunion-field-label'
+					onChange={ this.onChangeLabel }
+				/>
+				{ this.props.required && <span className="required">{ __( '(required)' ) }</span> }
```
Translation is missing `jetpack` textdomain.

georgestephanis replied:
Yeah, I'm waiting until the end to do a full pass adding it in everywhere.  Also, I understand there's some confusion about what the proper way to handle i18n strings is anyways, so may change if that shakes up.

georgestephanis replied:
Just added it in.

simison commented on modules/contact-form/js/gutenblock.jsx:
```diff
@@ -0,0 +1,630 @@
+		this.onChangeLabel = this.onChangeLabel.bind( this );
+	}
+
+	onChangeLabel( x ) {
```
`event` instead of `x` would seem better suited name for this variable?

georgestephanis replied:
Probably.  Was a leftover from it being a shorthand one liner.

georgestephanis replied:
Fixed.

simison commented on modules/contact-form/js/gutenblock.jsx:
```diff
@@ -0,0 +1,630 @@
+						label={ props.label }
+						setAttributes={ props.setAttributes }
+					/> }
+					disabled={ true }
```
I believe `disabled` would be the same as `disabled={ true }` here

georgestephanis replied:
Done

simison commented on modules/contact-form/js/gutenblock.jsx:
```diff
@@ -0,0 +1,630 @@
+
+	onChangeSubject( x ) {
+		this.props.setAttributes( { subject: x } );
+	}
```
We could shorten this a bit like so:
```js
onChangeSubject( subject ) {
    this.props.setAttributes( { subject } );
}
```

What do you think?

georgestephanis replied:
Sure.  I'm just unfamiliar with that new array syntax.  I'll get it done.

simison replied:
💪 you'll love shorthands once you get familiar with them, they're awesome. :-)

simison commented on modules/contact-form/js/gutenblock.jsx:
```diff
@@ -0,0 +1,630 @@
+					{this.props.children}
+					<TextControl
+						className="button button-primary button-default grunion-submit-button"
+						value={ this.props.submit_button_text }
```
Nit; this should be camelCased

georgestephanis replied:
This is intentionally not camelCased, as it's an attribute drawn from the legacy shortcode and it simplifies migrations by using the same attribute names for the shortcode and block.

simison replied:
Ah! Interesting. What would be a reason not to rename the attribute name in the shortcode transform function? I.e. how does it simplify the process?

georgestephanis replied:
By leaving the variable name the same, we both make the transformation process from shortcode to block simpler, as well as retain the ability to use a common interpreter for the front-end to render the form, whether it's a shortcode or a block.

simison commented on modules/contact-form/js/gutenblock.jsx:
```diff
@@ -0,0 +1,630 @@
+		},
+		submit_button_text: {
+			type: 'string',
+			'default': __( 'Submit »' )
```
Would _"Send"_ work as a less technical word here?

georgestephanis replied:
It's the legacy string, and I'd rather leave it in place so as to not unnecessarily churn translations.

georgestephanis committed:
Simplify the disabled attribute so we don't need the `true` assignment

georgestephanis committed:
Update some naming and patterns

Props @simison for the suggestions.

georgestephanis committed:
Add missing textdomains.

georgestephanis committed:
Use the Gutenberg generated className as well.

https://github.com/Automattic/jetpack/pull/10256#pullrequestreview-165139449

georgestephanis committed:
Better scope some CSS

georgestephanis committed:
Build.

georgestephanis committed:
Rearrange some icons and tidy some whitespace.

georgestephanis committed:
Build.

georgestephanis committed:
Renamespace everything from Grunion to Jetpack

georgestephanis committed:
Build.

georgestephanis commented:
I've also now stripped out the `grunion` phrase in favor of namespacing as Jetpack

georgestephanis committed:
First attempt at extracting to sub file

cc: @coderkevin

georgestephanis committed:
Minor bookkeeping

georgestephanis committed:
Build.

georgestephanis committed:
Extract out components for maintainability.

Co-Authored-By: Kevin Killingsworth <kevin@redfenix.com>

georgestephanis committed:
Build.

coderkevin reviewed (commented):
This is looking great, George. The file structure is nice and looks good as well. I'd approve if I felt like I had more experience with code in this repo. LGTM

georgestephanis committed:
Make scss version of gutenblock css

currently just a pull over.

simison commented on modules/contact-form/css/gutenblock.scss:
```diff
@@ -0,0 +1,84 @@
+
+.jetpack-field-label.jetpack-field-label.jetpack-field-label:focus,
+.jetpack-field-label.jetpack-field-label.jetpack-field-label:hover {
+	border-color: #8d96a0;
```
We don't have a mechanism to use Gutenberg colour variables directly yet, but let's define them in this file and use colours via variables like we do all across in Calypso.

See example: https://github.com/Automattic/wp-calypso/blob/a5c963f19173773d1732b3f23e2a60b961faa3a2/client/gutenberg/extensions/markdown/editor.scss

hypest commented:
👋. May I ask for a sample html snippet with the block serialised? (Like, a sample output of the `save()` function with the shortest/simplest block configuration that is valid). Thanks in advance!

georgestephanis committed:
Shuffle around files to match Calypso structure.

georgestephanis committed:
Build.

georgestephanis committed:
These changes are no longer necessary.

georgestephanis commented:
@hypest Depends what you mean by 'valid'

georgestephanis committed:
Make sure we're building the scss as well.

georgestephanis committed:
Build.

georgestephanis commented:
Build command is updated.

georgestephanis committed:
Convert the css to scss and build back to css.

georgestephanis committed:
Add RTL

georgestephanis committed:
Don't scan these files -- they're synced with Calypso and actually jsx or built files.

gwwar commented:
Is this far along enough to start manual testing? We can maybe have @Automattic/flowpatrol help with finding corner cases and breaking things ✨

kraftbj commented:
Obviously not required for this PR, but it made me realize we had RTL build tooling in gulp. It would be handy to start relying on it more with things like this so to avoid needing to do too much manual updating of styling in the future.

mapk committed:
Style revised to match default G blocks on elements for hover and focus states.

simison commented:
>RTL build tooling in gulp. It would be handy to start relying on it more with things like this so to avoid needing to do too much manual updating of styling in the future.

We have RTL support it in the Muriel SDK which will be the tool to build assets for this block eventually. :thumbsup: SDK uses the same library as Jetpack's build pipeline.

georgestephanis committed:
Rebuild editor.css

mapk committed:
Minor styling for the multiple checkbox.

georgestephanis committed:
Add a conditional for how the innerblocks are passed back up.

This is based off some local hacks I've made to Gutenberg -- as noted in  the following diff:

```
diff --git a/lib/blocks.php b/lib/blocks.php
index 3cd1ce33d..c3fad35a1 100644
--- a/lib/blocks.php
+++ b/lib/blocks.php
@@ -200,17 +200,33 @@ function _recurse_blocks( $blocks, $all_blocks ) {

 	foreach ( $blocks as $block ) {
 		$block = (array) $block;
+
+		$innerBlocksRendered = null;
+		if ( $block['innerBlocks'] ) {
+			$innerBlocksRendered = _recurse_blocks( $block['innerBlocks'], $all_blocks );
+		}
+
+		$block_content = $block['innerHTML'];
+		if ( $innerBlocksRendered ) {
+			$block_content = array(
+				'innerBlocksRendered' => $innerBlocksRendered,
+				'innerBlocksRaw'      => $block['innerBlocks'],
+				'innerHTML'           => $block['innerHTML'],
+			);
+		}
+
 		if ( in_array( $block['blockName'], $dynamic_blocks ) ) {
 			// Find registered block type. We can assume it exists since we use the
 			// `get_dynamic_block_names` function as a source for pattern matching.
 			$block_type = WP_Block_Type_Registry::get_instance()->get_registered( $block['blockName'] );

 			// Replace dynamic block with server-rendered output.
-			$block_content = $block_type->render( (array) $block['attrs'], $block['innerHTML'] );
-		} elseif ( $block['innerBlocks'] ) {
-			$block_content = _recurse_blocks( $block['innerBlocks'], $all_blocks );
-		} else {
-			$block_content = $block['innerHTML'];
+			$block_content = $block_type->render( (array) $block['attrs'], $block_content );
+		}
+
+		// If it was innerBlocks, but not a dynamic block.
+		if ( is_array( $block_content ) ) {
+			$block_content = $block_content['innerBlocksRendered'];
 		}

 		/**
```

georgestephanis committed:
Merge branch 'gm18/grunionblock' of https://github.com/Automattic/jetpack into gm18/grunionblock

georgestephanis commented:
Confirmed that once Gutenberg sorts out how it can pass innerBlocks to a render_callback the form will display beautifully.

mapk committed:
More form styling for the checkbox and radio options. Made checkbox for single and multiple match in size and color. Also removed the arrow from the 'submit' btn.

mapk committed:
Merge branch 'gm18/grunionblock' of github.com:Automattic/jetpack into gm18/grunionblock

mapk committed:
Removed arrow from form submit btn

mapk commented:
I'd also love to get the submit button to use the same styles as the Gutenberg default button except with Jetpack green.

![btn](https://user-images.githubusercontent.com/617986/47190018-3baf5680-d2f4-11e8-8115-01601a7a0f54.png)

sirreal commented:
I've been keeping an eye on this PR, but it sounds like it still has blockers so I've been holding off.

It seems that other folks have left sufficient initial feedback.

@georgestephanis if you have specific questions @Automattic/luna can help with at this time, just say so. Otherwise, I plan to wait until you consider this ready to merge. At that point, I'd prefer to review independent PRs in Calypso and Jetpack for the changes that are intended to be merged in each repository.

MichaelArestad commented:
> I'd also love to get the submit button to use the same styles as the Gutenberg default button except with Jetpack green.

@mapk This is a tough one. Normally, I would say go for it, but we need the front end styles to match the gutenberg styles. It's also going to be backwards compatible meaning for sites with a Contact form, if they update and get the Gutenberg block version, it should probably be the same. Does that make sense? 

If there was time, I would make the old styles a "style option" and make the default for new contact forms use the green button.

MichaelArestad commented:
I did a walkthrough and found these issues:

- [ ] It can be trick to select the Contact form container block. If I click at the top or bottom a form field block, I can select the Form block. However, it's unclear where to click. Core Gutenberg should already handle this elegantly somehow (I hope). @mtias ideas?
- [ ] Editing labels for a form field (Like "name"), the text input operates differently than the text inputs in Gutenberg, When focused and hovered, it should retain the focus style, but doesn't. The border also doesn't magically disappear when typing, but I'm okay with that for v1.
![image](https://user-images.githubusercontent.com/1123119/47245697-96f64d00-d3b7-11e8-8f2e-c1cbd12bf248.png)

- [ ] The submit button could use some top margin:
![image](https://user-images.githubusercontent.com/1123119/47245682-8a71f480-d3b7-11e8-9f78-fd900f14be6f.png)

- [ ] Are the inner blocks supposed to go all the way to the edge of the parent block? The inner block and insert block go all the way to the edge and it sort of breaks the block up visually. I think there should be padding between them.
![image](https://user-images.githubusercontent.com/1123119/47245768-dd4bac00-d3b7-11e8-9654-6c8100b8f090.png)

- [ ] This might be a core Gutenberg issue, but the inserter for innerblocks is a bit odd. It has the main category and then a couple under that including "Most used" and "Common blocks." This seems rather redundant. It would maaaaybe be helpful if broken into some categories like "text inputs" "radios" "checkboxes," but I think just having the first section would be better.

![image](https://user-images.githubusercontent.com/1123119/47245844-28fe5580-d3b8-11e8-8757-ccbf320cac27.png)
![image](https://user-images.githubusercontent.com/1123119/47245856-33b8ea80-d3b8-11e8-8671-54515772b6ae.png)
![image](https://user-images.githubusercontent.com/1123119/47245865-3b788f00-d3b8-11e8-8877-382f50ad8e1a.png)

- [ ] The inputs for the labels should have placeholders instead of text when adding a new one. For example, if I delete the text in a label, it should say "Add label" using input placeholder text instead of actual text. It shouldn't have "Type here..." in pre-filled text. Another thing that would be good when dropping in a new field is to have the label pre-filled with the name of the input. "Name" "Email" "URL"
![image](https://user-images.githubusercontent.com/1123119/47246032-f6089180-d3b8-11e8-8d1f-52e05fa2d01f.png)

- [ ] I couldn't get drag and drop to work for inner blocks.
- [ ] There are some lingering focus styles on the submit button when focused:
![image](https://user-images.githubusercontent.com/1123119/47246125-6dd6bc00-d3b9-11e8-9559-48e7794112ec.png)

- [ ] I think we can probably remove this from the Inspector since the button itself is editable:
![image](https://user-images.githubusercontent.com/1123119/47246151-8941c700-d3b9-11e8-8916-5925727e2e3b.png)


### Design exploration ideas:
* I would move the Required toggle next to the label somehow removing it from the inspector completely.
* Perhaps one or both of these fields should be in the block itself when selected. I would put it at the top. Instead of using normal inputs for these, perhaps you could make the subject line look like an h3 or something?
![image](https://user-images.githubusercontent.com/1123119/47246203-cd34cc00-d3b9-11e8-9f90-c596e73f4664.png)
 
What do you think @mapk >

mapk commented:
>It can be trick to select the Contact form container block. If I click at the top or bottom a form field block, I can select the Form block. However, it's unclear where to click. Core Gutenberg should already handle this elegantly somehow (I hope).

I created a screenshot showing all the hover borders of the blocks compared to the "columns" block in Gutenberg.

![outlines](https://user-images.githubusercontent.com/617986/47248623-81395580-d3c0-11e8-9b4b-76105fe6e116.png)


The horizontal spacing between the parent block and inner blocks appears correct. The only problem is the spacing between the parent block and the top inner block, and the spacing between the parent block and the bottom inner block. That spacing should match that of the horizontal spacing as well like the "columns" block displayed below in the screenshot.

mapk committed:
Styled the appender to hide with the block is not selected. Added some styling to it so that it reflects the image gallery appender a bit more.

mapk commented:
I styled the block appender for the form to look closer to the block appender in the Gutenberg core image gallery.

**Contact form appender**

![inserter](https://user-images.githubusercontent.com/617986/47308021-524ff900-d5e5-11e8-9246-c1720841b108.gif)

**Gutenberg core image gallery appender**

![image-inserter](https://user-images.githubusercontent.com/617986/47308035-5e3bbb00-d5e5-11e8-8574-2f31180cafa6.gif)

I am a bit confused b/c some of the styling from Gutenberg just shows a white block appender with a dashed line. But the block appender for galleries is a light gray box with solid lines. So I kept the dashed line but added a light gray background. On hover the dashed line turns to a solid line like the image gallery appender.

I also `display: none;` the appender until the main block was selected or any of the other inner blocks. This works similarly to the Gutenberg core image gallery appender as well.

georgestephanis committed:
Build.

georgestephanis committed:
Change "Type here..." to a placeholder instead of value.

georgestephanis committed:
Tidy up some missing options

georgestephanis committed:
Default the label value to the field type.

georgestephanis committed:
Build.

georgestephanis commented:
Current calypso pr: https://github.com/Automattic/wp-calypso/pull/27996

georgestephanis committed:
Switch checkbox and multiple components to use `BaseControl`

This will result in more consistent markup and styling.

Co-Authored-By: Mark Uraine <mapk@users.noreply.github.com>

georgestephanis committed:
Populate multis with the actual check and radio boxes

Co-Authored-By: Mark Uraine <mapk@users.noreply.github.com>

georgestephanis committed:
Build.

georgestephanis committed:
Switch submit button to ContentEditable instead.

Co-Authored-By: Mark Uraine <mapk@users.noreply.github.com>

georgestephanis committed:
Spaces to tabs

georgestephanis committed:
Build.

georgestephanis committed:
Bump this around for scss structure nesting.

georgestephanis committed:
Build.

georgestephanis committed:
Fix the case of the vanishing appender.

Sometimes when you're typing in an input field, the .is-selected goes away, but it still keeps .is-typing -- so let's use the general sibling selector on that to find our way to the appender we want to display.

georgestephanis committed:
Build.

georgestephanis committed:
suppressContentEditableWarning

georgestephanis committed:
Build.

mapk committed:
Fixed minor bug with alignment of 'x' btn for multiple radios and checkboxes. Also styled the single checkbox to look like the new disabled checkboxes.

mapk committed:
Merge branch 'gm18/grunionblock' of github.com:Automattic/jetpack into gm18/grunionblock

georgestephanis committed:
Reorganize scss for better hierarchy and make submit button not contentEditable

contentEditable seems to introduce way more bugs than an input field ever had.  For v1 let's only allow it to be changed via the Inspector.

georgestephanis committed:
Build.

georgestephanis committed:
Add in a new caveat for block rendering.

georgestephanis committed:
Oops, the type was getting lost on these fields.  Better add in A Better Way™️

georgestephanis committed:
TABS TABS TABS

georgestephanis committed:
Correcting the default fieldset to match historical configurations.

georgestephanis committed:
Whitespace

georgestephanis committed:
Allow all blocks inside a Contact Form.

georgestephanis committed:
Build.

georgestephanis committed:
Clear the last vestiges of `»` away from the submit button.

Co-Authored-By: Mark Uraine <mapk@users.noreply.github.com>

mapk committed:
Fixed submit btn padding

mapk commented:
### Inner Block Picker

1. Only show the "Contact Form" category with all 11 inner blocks (form fields).
2. Order the inner blocks most likely to be used with logical groupings (ie. text fields, checkboxes, radio/select, etc.)
3. Let's update some of these icons as well with Material's icons.

**Example of Inner Block Picker (ordered alphabetically)**

![picker-inner-block](https://user-images.githubusercontent.com/617986/47461941-e0b7ac80-d796-11e8-8856-494afc88fcf7.png)


**Icons**
Checkbox = https://material.io/tools/icons/?icon=check_box&style=outline
Checkbox Multiple = https://material.io/tools/icons/?icon=ballot&style=outline
Date = https://material.io/tools/icons/?icon=today&style=outline
Email = https://material.io/tools/icons/?icon=email&style=outline
Name = https://material.io/tools/icons/?icon=person&style=outline
Radio = https://material.io/tools/icons/?icon=radio_button_checked&style=outline
Select = https://material.io/tools/icons/?icon=calendar_view_day&style=outline
Telephone = https://material.io/tools/icons/?icon=phone&style=outline
Text = https://material.io/tools/icons/?icon=short_text&style=outline
Textarea = https://material.io/tools/icons/?icon=notes&style=outline
URL = https://material.io/tools/icons/?icon=computer&style=outline

@jasmussen any thoughts on the ordering of inner blocks in the picker?
@MichaelArestad @georgestephanis Sound good?

jasmussen commented:
> any thoughts on the ordering of inner blocks in the picker?

I think it's alphabetical so I'm not sure there's much you can do.

georgestephanis committed:
Disallow non-field child blocks for now.

We may whitelist select blocks down the road.

Co-Authored-By: Mark Uraine <mapk@users.noreply.github.com>

georgestephanis committed:
Build.

georgestephanis committed:
Swap in most SVGs from Material Design for fields.

georgestephanis committed:
Build.

georgestephanis committed:
Swap to the specified outline style icons.

georgestephanis committed:
Switch over to Gutenberg's SVG primitives for accessibility.

https://github.com/WordPress/gutenberg/tree/5c2085ada83688b441b5d784ad35ee9230759f77/packages/components/src/primitives/svg

georgestephanis committed:
Build.

georgestephanis commented:
Icons are swapped over to the specified material design ones, but I'm not sure how I can hide the other block selector tabs --

![screen shot 2018-10-25 at 3 18 32 pm](https://user-images.githubusercontent.com/941023/47524818-a79c3c80-d869-11e8-91c7-475ea6463b49.png)

georgestephanis commented:
If we can get https://github.com/WordPress/gutenberg/pull/11082 or the like merged to Gutenberg, this should be good to go.

georgestephanis committed:
Strip out legacy ways of rendering -- no longer useful.

catehstn commented:
@georgestephanis #11334 has now been merged - can we get this one in, now?

lezama commented:
@catehstn @Automattic/poseidon is working on it, we should have something by the end of the day.

lezama commented:
This has been merged in https://github.com/Automattic/wp-calypso/pull/27996.

Thanks @georgestephanis !

lezama commented:
In order to test the new Contact Form Block please follow the instructions on https://github.com/Automattic/jetpack/pull/10537

erikdemarco commented:
how to insert contactform gutenblock programatically from my functions.php?
